src/eval.ml

```diff
@@ -165,6 +169,91 @@ and eq_modulo : term -> term -> bool = fun a b ->
   let res = try eq_modulo [(a,b)]; true with Exit -> false in
   if !log_enabled then log_eqmd (r_or_g res "%a == %a") pp a pp b; res
 
+(** [tree_walk t s] tries to match stack [s] against tree [t] *)
+and tree_walk : Dtree.t -> stack -> (term * stack) option = fun itree istk ->
+  (* Count maximum number of stored items to initialize variable array *)
+  let capacity = Dtree.iter ~do_leaf:(fun _ _ -> 0)
+    ~do_node:(fun _ store subr abstr defr ->
+      let maxch = if subr = [] then 0
+        else List.extremum (>) (snd (List.split subr)) in
+      let maxchd = match defr with
+        | None     -> maxch
+        | Some(dr) -> max dr maxch in
+      let maxchdab = match abstr with
+        | None        -> maxchd
+        | Some(_, ar) -> max ar maxchd in
+      if store then succ maxchdab else maxchdab)
+    ~fail:0 itree in
+  let vars = Array.init capacity (fun _ -> Patt(None, "", [| |])) in
+
+  (* [walk t s c] where [s] is the stack of terms to match and [c] the cursor
+     indicating where to write in the [env] array . *)
```

Comment:
What is [env]?

Reply:
The environment of an rhs containing terms to be bound; the item slots of patterns point to.